client/client-lib/src/mint/mod.rs

```diff
@@ -70,121 +71,60 @@ impl<'c> MintClient<'c> {
-        let coins = self.select_and_spend_coins(batch.subtransaction(), amount)?;
-        let (spend_keys, coins) = self.create_coin_input_from_coins(coins)?;
-        batch.commit();
-        Ok((spend_keys, coins))
-    }
-
-    pub fn create_coin_input_from_coins(
-        &self,
-        coins: Coins<SpendableCoin>,
-    ) -> Result<(Vec<KeyPair>, Coins<Coin>)> {
-        let coin_key_pairs = coins
-            .into_iter()
-            .map(|(amt, coin)| {
-                let spend_key =
-                    bitcoin::KeyPair::from_seckey_slice(self.context.secp, &coin.spend_key)
-                        .map_err(|_| MintClientError::ReceivedUspendableCoin)?;
-
-                // We check for coin validity in case we got it from an untrusted third party. We
-                // don't want to needlessly create invalid tx and bother the federation with them.
-                let spend_pub_key = spend_key.public_key();
-                if &spend_pub_key == coin.coin.spend_key() {
-                    Ok((spend_key, (amt, coin.coin)))
-                } else {
-                    Err(MintClientError::ReceivedUspendableCoin)
-                }
-            })
-            .collect::<Result<Vec<_>>>()?;
-        Ok(coin_key_pairs.into_iter().unzip())
-    }
-
-    /// Adds change to a TransactionBuilder as the last output (if necessary) and builds the final Transaction
-    pub fn finalize_change<R: Rng + CryptoRng>(
+    pub async fn submit_tx_with_change<R: RngCore + CryptoRng>(
         &self,
-        change_required: Amount,
-        batch: BatchTx,
-        mut tx: TransactionBuilder,
-        mut rng: R,
-    ) -> Transaction {
-        let min_change = *self.context.config.tbs_pks.tiers().min().unwrap();
-
-        if change_required >= min_change {
-            self.create_coin_output(batch, change_required, &mut rng, |change_output| {
-                tx.output(Output::Mint(change_output));
-                OutPoint {
-                    txid: tx.tx.tx_hash(),
-                    out_idx: (tx.tx.outputs.len() - 1) as u64,
-                }
-            });
-        }
+        fee_consensus: &FeeConsensus,
+        tx: TransactionBuilder,
+        mut batch: Accumulator<BatchItem>,
+        rng: R,
+    ) -> Result<TransactionId> {
+        let change_required = tx.change_required(fee_consensus);
+        let final_tx = tx.build(
+            change_required,
+            batch.transaction(),
+            self.context.secp,
+            &self.context.config.tbs_pks,
+            rng,
+        );
+        let txid = final_tx.tx_hash();
+        let mint_tx_id = self.context.api.submit_transaction(final_tx).await?;
+        // TODO: make check part of submit_transaction
+        assert_eq!(
+            txid, mint_tx_id,
+            "Federation is faulty, returned wrong tx id."
+        );
 
-        tx.build(self.context.secp, &mut rng)
+        self.context.db.apply_batch(batch).expect("DB error");
```

Comment:
From a crash-safety pov we should first save and then submit imo. But that's a bigger issue (#35) and could be a separate PR, just something we need to keep in mind.

Reply:
I think what we want is:

1. "Stage" tx in DB (this PR introduces this intermediate state)
2. Submit tx
3. If an error occurs either on submit or the tx isn't confirmed for some reason we can later rollback the staged tx in the client DB. If the tx succeeds we finalize the staged tx in the client DB.

I assumed this `apply_batch` invocation was for step 1, but maybe I'm mistaken?